src/main/groovy/com/bmuschko/gradle/docker/DockerConventionJvmApplicationPlugin.groovy

```diff
@@ -67,16 +68,21 @@ abstract class DockerConventionJvmApplicationPlugin<EXT extends DockerConvention
         EXT extension = configureExtension(project.objects, dockerExtension)
 
         project.plugins.withType(JavaPlugin) {
-            Dockerfile createDockerfileTask = createDockerfileTask(project, extension)
-            Sync syncBuildContextTask = createSyncBuildContextTask(project, createDockerfileTask)
-            createDockerfileTask.dependsOn syncBuildContextTask
-            DockerBuildImage dockerBuildImageTask = createBuildImageTask(project, createDockerfileTask, extension)
-            createPushImageTask(project, dockerBuildImageTask)
+            TaskProvider<Dockerfile> createDockerfileTask = registerDockerfileTask(project, extension)
+            TaskProvider<Sync> syncBuildContextTask = registerSyncBuildContextTask(project, createDockerfileTask)
+            createDockerfileTask.configure(new Action<Dockerfile>() {
```

Comment:
I can look at this myself but it somewhat looks like a circular dependency between these tasks of some sort.

Reply:
Yeah it does look like that. I debugged it and stepped through the code.

Here's what happens when `dockerCreateDockerfile` task is executed
1. During the configuration phase, the `dockerCreateDockerfile`
2. During configuration, `dockerCreateDockerfile` tasks is configured to depend on `dockerSyncBuildContext`. However, the dependency is done via `TaskProvider`, so when the dependency is declared, the `dockerSyncBuildContext` has not yet been configured
3. The `dockerSyncBuildContext` is configured
4. The `dockerCreateDockerfile.get()` call just returns the configured task (which finished in step 3)

I'm open for suggestions to how improve this, but it does appear to work just fine